### PR TITLE
3167 UI credentials collectors

### PR DESCRIPTION
**Pull request by ilija-lazoroski**
# What does this PR do?

Fixes part of #3167 .

Things left to do:
- [x] Fix credentials collectors ui readonly if plugin not selected
- [x] Fix BB test configurations that adds credentials collectors

## PR Checklist
* [x] Have you added an explanation of what your changes do and why you'd like to include them?
* [ ] Is the TravisCI build passing?
* [x] Was the CHANGELOG.md updated to reflect the changes?
* [x] Was the documentation framework updated to reflect the changes?
* [x] Have you checked that you haven't introduced any duplicate code?

## Testing Checklist

* [x] Added relevant unit tests?
* [x] Do all unit tests pass?
* [x] Do all end-to-end tests pass?
* [x] Any other testing performed?
    > Tested by {Running the Monkey locally with relevant config/running Island/...}
* [x] If applicable, add screenshots or log transcripts of the feature working

https://user-images.githubusercontent.com/15820737/234634593-98dd0bb5-1366-49d0-8d0e-06942ba3375b.mov

### CODE REVIEW

File: monkey/common/agent_configuration/agent_configuration.py [hunk 1]
```diff
@@ -15,7 +15,7 @@ class AgentConfiguration(MutableInfectionMonkeyBaseModel):
         "seconds)",
         default=30,
     )
-    credentials_collectors: Tuple[PluginConfiguration, ...] = Field(
+    credentials_collectors: Optional[CredentialsCollectorsConfiguration] = Field(
```
Comment (mssalvatore) [inline monkey/common/agent_configuration/agent_configuration.py:18]: Make this match `ExploitationConfiguration.exploiters`. It shouldn't be optional and we don't need to define a separate configuration.

File: monkey/monkey_island/cc/ui/src/components/configuration-components/FormConfig.tsx [hunk 2]
```diff
@@ -0,0 +1,43 @@
+import Form from '@rjsf/bootstrap-4';
+import React, {useState} from 'react';
+import _ from 'lodash';
+
+
+export const EXPLOITERS_PATH_PROPAGATION = 'exploitation.exploiters';
+export const EXPLOITERS_CONFIG_PATH = 'propagation.' + EXPLOITERS_PATH_PROPAGATION;
```
Comment (shreyamalviya) [inline monkey/monkey_island/cc/ui/src/components/configuration-components/FormConfig.tsx:7]: Are these constants needed here? Let's move them to where they're relevant.

Comment (ilija-lazoroski) [reply monkey/monkey_island/cc/ui/src/components/configuration-components/FormConfig.tsx:7]: Probably just a leftover. I will re/move them as needed. Thanks!

File: monkey/common/agent_configuration/agent_configuration.py [hunk 3]
```diff
@@ -15,9 +15,11 @@ class AgentConfiguration(MutableInfectionMonkeyBaseModel):
...
         default=30,
     )
-    credentials_collectors: Tuple[PluginConfiguration, ...] = Field(
-        title="Credentials collectors",
-        description="Configure options for the attack’s credentials collection stage",
+    credentials_collectors: Dict[str, Dict] = Field(
+        title="Enabled credentials collectors",
+        description="Click on an credentials collectors to get more information"
+        " about it. \n \u26A0 Note that using unsafe credentials collectors may"
+        " cause crashes of the local machine/system.",
```
Comment (shreyamalviya) [inline monkey/common/agent_configuration/agent_configuration.py:22]:
```suggestion
        description="Click on a credentials collectors to get more information"
        " about it. \n \u26A0 Note that using unsafe options may"
        " result in unexpected behavior on the machine.",
```

File: monkey/monkey_island/cc/ui/src/components/configuration-components/FormConfig.tsx [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+  const setUiSchemaForCurrentSection = (uiSubschema, path) => {
+    let newSchema = _.cloneDeep(formUiSchema);
+    if(path !== selectedSection){
+      _.set(newSchema, path, uiSubschema);
+      setFormUiSchema(newSchema);
+    }else {
+      setFormUiSchema(uiSubschema);
+    }
+  }
```
Comment (shreyamalviya) [inline monkey/monkey_island/cc/ui/src/components/configuration-components/FormConfig.tsx:24]:
```suggestion
    if (path !== selectedSection) {
      let newSchema = _.cloneDeep(formUiSchema);
      _.set(newSchema, path, uiSubschema);
      setFormUiSchema(newSchema);
    } else {
      setFormUiSchema(uiSubschema);
    }
  }
```

File: monkey/monkey_island/cc/ui/src/components/configuration-components/FormConfig.tsx [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+                   'setSelectedPlugins': setSelectedPlugins,
+                   'setUiSchema': setUiSchemaForCurrentSection,
+                   'section': selectedSection
+                 }}>
+                 <button type='submit' className={'hidden'}>Submit</button>
+             </Form>
+    }
+  return (<div>
+    {getForm()}
+  </div>)
```
Comment (shreyamalviya) [inline monkey/monkey_island/cc/ui/src/components/configuration-components/FormConfig.tsx:39]: Is this function necessary? Just do:
```suggestion
return (
    <Form {...props}
          uiSchema={formUiSchema}
          formContext={{'selectedPlugins': selectedPlugins,
                        'setSelectedPlugins': setSelectedPlugins,
                        'setUiSchema': setUiSchemaForCurrentSection,
                        'section': selectedSection
                      }}>
      <button type='submit' className={'hidden'}>Submit</button>
    </Form>
  )
```

File: monkey/monkey_island/cc/ui/src/components/configuration-components/PluginSelectorTemplate.tsx [hunk 6]
```diff
@@ -8,12 +8,14 @@ import {InfoPane, WarningType} from '../ui-components/InfoPane';
 import {EXPLOITERS_PATH_PROPAGATION} from './PropagationConfig';
 import MarkdownDescriptionTemplate from './MarkdownDescriptionTemplate';
 
+const pluginSchemaPath = {'propagation': EXPLOITERS_PATH_PROPAGATION, 'credentials_collectors': 'credentials_collectors'}
+
```
Comment (shreyamalviya) [inline monkey/monkey_island/cc/ui/src/components/configuration-components/PluginSelectorTemplate.tsx:12]: This is inconsistent. If we have constants for the exploiters configuration path in the file relevant to them, we should have them for credential collectors too.

File: monkey/monkey_island/cc/ui/src/components/pages/ConfigurePage.js [hunk 7]
```diff
@@ -38,6 +38,7 @@ const CONFIGURED_PROPAGATION_CREDENTIALS_URL = '/api/propagation-credentials/con
 const configSubmitAction = 'config-submit';
 const configExportAction = 'config-export';
 const configSaveAction = 'config-saved';
+const CREDENTIALS_COLLECTORS_CONFIG_PATH = 'credentials_collectors';
```
Comment (shreyamalviya) [inline monkey/monkey_island/cc/ui/src/components/pages/ConfigurePage.js:41]: This is related to the above comment. Move this constant to the relevant file and import it like we've done for exploiters.

File: monkey/monkey_island/cc/ui/src/components/utils/SafeOptionValidator.js [hunk 8]
```diff
@@ -19,6 +14,11 @@ function getPluginDescriptors(schema, config) {
       name: 'Exploiters',
       allPlugins: schema.properties.propagation.properties.exploitation.properties.exploiters.properties,
       selectedPlugins: Object.keys(config.propagation.exploitation.exploiters)
+    },
+    {
+      name: 'CredentialsCollectors',
+      allPlugins: schema.properties.credentials_collectors.properties,
+      selectedPlugins: Object.keys(config.credentials_collectors)
```
Comment (shreyamalviya) [inline monkey/monkey_island/cc/ui/src/components/utils/SafeOptionValidator.js:21]: If we change the configuration path of something, we'll need to update to here as well as `PluginSelectorTemplate.tsx`.

I see that the way we access the configuration is different in both places, but we should have the paths at one place and create a utility method that accepts the path as a parameter which we can use in this file. Maybe in a different PR.

**Review by shreyamalviya — CHANGES_REQUESTED** [review]

**Review by mssalvatore — APPROVED** [review]

**Review by shreyamalviya — APPROVED** [review]

**Pull request merged**